### PR TITLE
Making every type optional via Option

### PR DESCRIPTION
**Pull request by 2022H1030042G**
Refer: #4399 

## Changelog

- [ ] The PR requires a changelog line

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

File: crates/rome_service/src/configuration/formatter.rs [hunk 1]
```diff
@@ -9,24 +9,24 @@ use serde::{Deserialize, Serialize};
 #[serde(rename_all = "camelCase", default, deny_unknown_fields)]
 pub struct FormatterConfiguration {
     // if `false`, it disables the feature. `true` by default
-    pub enabled: bool,
+    pub enabled: Option<bool>,
```
Comment (ematipico) [inline crates/rome_service/src/configuration/formatter.rs:12]: For each line where you added `Option`, you add the following line, for example:

```rs
#[serde(skip_serializing_if = "Option::is_none")]
pub enabled: Option<bool>,
```

File: crates/rome_service/src/configuration/parse/json/formatter.rs [hunk 2]
```diff
@@ -27,24 +27,24 @@ impl VisitNode<JsonLanguage> for FormatterConfiguration {
...
             }
             "indentSize" => {
-                self.indent_size = self.map_to_u8(&value, name_text, u8::MAX, diagnostics)?;
+                self.indent_size = self.map_to_u8(&value, name_text, u8::MAX, diagnostics);
             }
             "lineWidth" => {
-                let line_width = self.map_to_u16(&value, name_text, LineWidth::MAX, diagnostics)?;
+                let line_width = self.map_to_u16(&value, name_text, LineWidth::MAX, diagnostics);
 
                 self.line_width = match LineWidth::try_from(line_width) {
```
Comment (ematipico) [inline crates/rome_service/src/configuration/parse/json/formatter.rs:49]: You have to update this line too with `Some(match LineWith::try_from() {})`

File: crates/rome_service/src/configuration/formatter.rs (hunk 1 above)
Comment (2022H1030042G) [reply crates/rome_service/src/configuration/formatter.rs:12]: > For each line where you added `Option`, you add the following line, for example:
> 
> ```rust
> #[serde(skip_serializing_if = "Option::is_none")]
> pub enabled: Option<bool>,
> ```

This #[serde(skip_serializing_if = "Option::is_none")] has to be exactly added for all the lines?

File: crates/rome_service/src/configuration/parse/json/formatter.rs (hunk 2 above)
Comment (2022H1030042G) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: Do you mean line 47 has to be replaced with Some(match LineWith::try_from() {}) ?

File: crates/rome_service/src/configuration/formatter.rs (hunk 1 above)
Comment (ematipico) [reply crates/rome_service/src/configuration/formatter.rs:12]: Yes, for all the lines you added `Option`

File: crates/rome_service/src/configuration/parse/json/formatter.rs (hunk 2 above)
Comment (ematipico) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: No, line 49

Comment (2022H1030042G) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: error
it gives this error.

Comment (2022H1030042G) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: even if I remove ` , it gives the error:

error2

Comment (ematipico) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: Sorry, that was an example 😅 I didn't want to write all the code. You just need to take the existing code:

```rs
self.line_width = match LineWith::try_from() {/* existing code */}
```

To 

```rs
self.line_width = Some(match LineWith::try_from() {/* existing code */})
```

Comment (2022H1030042G) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: There are still about 18 errors

Comment (2022H1030042G) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: I did, how can I solve an error like this?: 
error3

Comment (ematipico) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: Look for a function called `serialize_line_width` and `deserialize_line_width`. These two functions have `LineWidth` in their signatures. Change them to `Option<LineWidth>`. You will have some compiler errors. The compiler will suggest you how to fix them.

Comment (2022H1030042G) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: for line 8 shown above, what can be done?

Comment (ematipico) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: Could you please push the latest changes? The error is probably somewhere else.

Comment (2022H1030042G) [reply crates/rome_service/src/configuration/parse/json/formatter.rs:49]: Sure

File: crates/rome_formatter/src/lib.rs [hunk 3]
```diff
@@ -83,7 +83,7 @@ pub enum IndentStyle {
     #[default]
     Tab,
     /// Space, with its quantity
-    Space(u8),
+    Space(Option<u8>),
```
Comment (ematipico) [inline crates/rome_formatter/src/lib.rs:86]: This doesn't need to change. You can revert the change

File: crates/rome_formatter/src/printer/printer_options/mod.rs [hunk 4]
```diff
@@ -74,7 +74,7 @@ impl PrinterOptions {
     pub(super) const fn indent_width(&self) -> u8 {
         match self.indent_style {
             IndentStyle::Tab => self.tab_width,
-            IndentStyle::Space(count) => count,
+            IndentStyle::Space(Option<count>) => count,
```
Comment (ematipico) [inline crates/rome_formatter/src/printer/printer_options/mod.rs:77]: You can revert the change here

File: crates/rome_formatter/src/format_element/document.rs [hunk 5]
```diff
@@ -153,7 +153,7 @@ struct IrFormatOptions;
 
 impl FormatOptions for IrFormatOptions {
     fn indent_style(&self) -> IndentStyle {
-        IndentStyle::Space(2)
+        IndentStyle::Space(Some(2))
```
Comment (ematipico) [inline crates/rome_formatter/src/format_element/document.rs:156]: You can revert the change

File: crates/rome_formatter/src/format_element/document.rs [hunk 6]
```diff
@@ -165,7 +165,7 @@ impl FormatOptions for IrFormatOptions {
             tab_width: 2,
             print_width: self.line_width().into(),
             line_ending: LineEnding::LineFeed,
-            indent_style: IndentStyle::Space(2),
+            indent_style: IndentStyle::Space(Some(2)),
```
Comment (ematipico) [inline crates/rome_formatter/src/format_element/document.rs:168]: You can revert the change

File: crates/rome_service/src/configuration/formatter.rs [hunk 7]
```diff
@@ -67,8 +72,8 @@ impl TryFrom<FormatterConfiguration> for FormatSettings {
 
     fn try_from(conf: FormatterConfiguration) -> Result<Self, Self::Error> {
         let indent_style = match conf.indent_style {
-            PlainIndentStyle::Tab => IndentStyle::Tab,
-            PlainIndentStyle::Space => IndentStyle::Space(conf.indent_size),
+            Some(PlainIndentStyle::Tab) => IndentStyle::Tab,
+            Some(PlainIndentStyle::Space) => IndentStyle::Space(conf.indent_size),
```
Comment (ematipico) [inline crates/rome_service/src/configuration/formatter.rs:76]:
```suggestion
            Some(PlainIndentStyle::Tab) => IndentStyle::Tab,
            Some(PlainIndentStyle::Space) => IndentStyle::Space(conf.indent_size),
            None => IndentStyle::Tab
```

File: crates/rome_service/src/configuration/formatter.rs [hunk 8]
```diff
@@ -97,15 +102,15 @@ impl TryFrom<FormatterConfiguration> for FormatSettings {
...
 }
 
-fn deserialize_line_width<'de, D>(deserializer: D) -> Result<LineWidth, D::Error>
+fn deserialize_line_width<'de, D>(deserializer: D) -> Result<Option<LineWidth>, D::Error>
 where
     D: serde::de::Deserializer<'de>,
 {
     let value: u16 = Deserialize::deserialize(deserializer)?;
-    LineWidth::try_from(value).map_err(serde::de::Error::custom)
+    Option::<LineWidth>::try_from(value).map_err(serde::de::Error::custom)
```
Comment (ematipico) [inline crates/rome_service/src/configuration/formatter.rs:110]:
```suggestion
    let line_width = LineWidth::try_from(value).map_err(serde::de::Error::custom)?;
	Ok(Some(line_width))
```

**Review by ematipico — CHANGES_REQUESTED** [review]

**Comment by 2022H1030042G** [comment]
After performing the above changes, there are still errors.

**Comment by ematipico** [comment]
> After performing the above changes, there are still errors. PFA [errors.txt](https://github.com/rome/tools/files/11333629/errors.txt)

You can push the changes :) no need to attach any files!

**Comment by 2022H1030042G** [comment]
> > After performing the above changes, there are still errors. PFA [errors.txt](https://github.com/rome/tools/files/11333629/errors.txt)
> 
> You can push the changes :) no need to attach any files!

Done.

**Comment by ematipico** [comment]
Hi @2022H1030042G, I thought you had some basic knowledge of Rust. I needed this feature for #4405 and eventually did it myself there.

Thank you for your contribution. I am going to create some new good first issues that don't require priority from our end!

**Pull request closed without merging**